fhir-bulkimportexport-webapp/src/main/java/com/ibm/fhir/jbatch/bulkdata/export/fast/ResourcePayloadReader.java

```diff
@@ -0,0 +1,683 @@
+        this.txEndTime = readStartTime + this.txTimeoutMillis * 1000000L;
+
+        // Note we're already running inside a transaction (started by the Javabatch framework)
+        // so this txn will just wrap it...the commit won't happen until the checkpoint
+        FHIRTransactionHelper txn = new FHIRTransactionHelper(fhirPersistence.getTransaction());
+        txn.begin();
+        boolean moreData = true;
+        try {
+            // Keep scanning while there is more data to fetch but stop the scan early
+            // so we don't exceed the transaction timeout. This break also allows a
+            // checkpoint in case we need to restart for any reason.
+            while (!isTxTimeExpired() && moreData) {
+
+                // Ask the persistence layer to start fetching the records in the current scan window,
+                // using the #processPayload lambda to handle each value as it is retrieved.
+                if (logger.isLoggable(Level.FINE)) {
+                    logger.fine("Fetching data from: " + fromLastModified + " to " + toLastModified);
+                }
+
+                // Make a note of how many resources were actually processed. If none were processed, this
+                // is also an indication that we've reached the end.
+                final int previouslyProcessed = this.resourcesProcessed;
+                ResourcePayload last = fhirPersistence.fetchResourcePayloads(resourceType, fromLastModified, toLastModified, rp->processPayload(rp));
+                final int newlyProcessed = this.resourcesProcessed - previouslyProcessed;
+
+                // check for errors which may have occurred during processing (rare)
+                if (this.processingException != null) {
+                    logger.log(Level.SEVERE, logPrefix(), this.processingException);
+                    throw this.processingException;
+                }
+
+                if (last != null && newlyProcessed > 0) {
+                    // Update our state so that we can start the next scan from the correct position.
+                    this.fromLastModified = last.getLastUpdated();
+                } else {
+                    // no data returned, so that's the end of the road
+                    logger.fine(logPrefix() + " no more data");
+                    moreData = false;
+                }
+            }
+
+            if (uploadId != null && !moreData) {
+                // There's no more data to fetch, so finalize the current upload before we leave
+                completeCurrentUpload();
+            }
+        } finally {
+            txn.end();
+
+            // Log a simple stat to give us an idea how quickly we processed the data
+            double delta = (System.nanoTime() - readStartTime) / 1e9;
+            logger.info(String.format("%s processed %d resources in %.2f seconds (rate=%.1f resources/second)", logPrefix(), this.resourcesProcessed, delta, resourcesProcessed/delta));
```

Comment:
minor style suggestion
```suggestion
            logger.info(String.format("%s processed %d resources in %.2f seconds (rate=%.1f resources/second)",
                    logPrefix(), this.resourcesProcessed, delta, resourcesProcessed/delta));
```